blockprod/src/detail/mod.rs

```diff
@@ -424,36 +446,36 @@ impl BlockProduction {
                 ));
             }
 
-            // TODO: see if we can simplify this
-            let transactions = match transactions_source.clone() {
-                TransactionsSource::Mempool => {
-                    // We conservatively use the minimum timestamp here in order to figure out
-                    // which transactions are valid for the block.
-                    // TODO: Alternatively, we can construct the transaction sequence from the
-                    // scratch every time a different timestamp is attempted. That is more costly
-                    // in terms of computational resources but will allow the node to include more
-                    // transactions since the passing time may release some time locks.
-                    let accumulator = self
-                        .collect_transactions(
-                            current_tip_index.block_id(),
-                            min_constructed_block_timestamp,
-                        )
-                        .await?;
-                    match accumulator {
-                        Some(acc) => acc.transactions().clone(),
-                        None => {
-                            // If the mempool rejects the accumulator (due to tip mismatch, or otherwise), we should start over and try again
-                            log::info!(
-                                "Mempool rejected the transaction accumulator. Restarting the block production attempt."
-                            );
-                            continue;
-                        }
+            // We conservatively use the minimum timestamp here in order to figure out
+            // which transactions are valid for the block.
+            // TODO: Alternatively, we can construct the transaction sequence from the
+            // scratch every time a different timestamp is attempted. That is more costly
+            // in terms of computational resources but will allow the node to include more
+            // transactions since the passing time may release some time locks.
+            let collected_transactions = {
+                let accumulator = self
+                    .collect_transactions(
+                        current_tip_index.block_id(),
+                        min_constructed_block_timestamp,
+                        transactions.clone(),
+                        transaction_ids.clone(),
+                        include_mempool,
+                    )
+                    .await?;
+
+                match accumulator {
+                    Some(acc) => acc.transactions().clone(),
+                    None => {
+                        // If the mempool rejects the accumulator (due
+                        // to tip mismatch, or otherwise), we should
+                        // start over and try again
+                        log::info!("Mempool rejected the transaction accumulator. Trying again.");
```

Comment:
I don't like this "trying again" compared to what was there before. It's not clear what the implication is. What happens is that the whole block production attempt is restarted, but this isn't clear here.

Reply:
This one is tricky because we're now combining both cases. I've made a change to throw away mempool transactions if the mempool comes back None. This way, we're kind of back to the original behavior of only using the supplied transactions.